lgc/state/PalMetadata.cpp

```diff
@@ -1263,6 +1263,45 @@ bool PalMetadata::fragmentShaderUsesMappedBuiltInInputs() {
   return false;
 }
 
+// =====================================================================================================================
+// Returns true if PS input and Pre-raster stage output can match and update SPI_PS_INPUT_CNTL.
+bool PalMetadata::matchInputOutputSemanticMetadata() {
+  auto inputSemanticIt = m_pipelineNode.find(Util::Abi::PipelineMetadataKey::PsInputSemantic);
+  if (inputSemanticIt == m_pipelineNode.end())
+    return true; // ignore pre-raster output semantic
+  auto outputSemanticIt = m_pipelineNode.find(Util::Abi::PipelineMetadataKey::PrerasterOutputSemantic);
+  if (outputSemanticIt == m_pipelineNode.end())
+    return false; // no pre-raster output semantic
+  // Find matched output semantic and update SPI_PS_INPUT_CNTL_*.offset
+  auto inputSemanticArray = inputSemanticIt->second.getArray(true);
+  auto outputSemanticArray = outputSemanticIt->second.getArray(true);
+
+  // Update SPI_PS_INPUT_CNTL_*.OFFSET
+  auto spiPsInputCntlArray = m_pipelineNode[Util::Abi::PipelineMetadataKey::GraphicsRegisters]
+                                 .getMap(true)[Util::Abi::GraphicsRegisterMetadataKey::SpiPsInputCntl]
+                                 .getArray(true);
+
+  unsigned elemIdx = 0;
+  unsigned matchedElemCount = 0;
+  for (auto inputSemantic : inputSemanticArray) {
+    const unsigned inSemantic = inputSemantic.getMap(true)[Util::Abi::PsInputSemanticMetadataKey::Semantic].getUInt();
+    auto spiPsInputCntElem = spiPsInputCntlArray[elemIdx].getMap(true);
+    for (auto outputSemantic : outputSemanticArray) {
+      const unsigned outSemantic =
+          outputSemantic.getMap(true)[Util::Abi::PrerasterOutputSemanticMetadataKey::Semantic].getUInt();
+      if (inSemantic == outSemantic) {
+        const unsigned mappedLoc =
+            outputSemantic.getMap(true)[Util::Abi::PrerasterOutputSemanticMetadataKey::Index].getUInt();
+        spiPsInputCntElem[Util::Abi::SpiPsInputCntlMetadataKey::Offset] = mappedLoc;
+        ++matchedElemCount;
+      }
+    }
```

Comment:
Presumably this loop should:

* Early-exit when the pre-raster output has been found
* Have an assertion that is triggered when no pre-raster output is found

Reply:
This function is removed as the match for semantic is handled by PAL. In the follow-up patch, we just need to loose the restriction, that is the check for `fragmentShaderUsesMappedBuiltInInputs()` is only required for relocatable shaders.